File: frontend/src/lib/seo/meta.test.ts
// SPDX-FileCopyrightText: 2026 Quad4 Software
// SPDX-License-Identifier: Apache-2.0

import { beforeEach, describe, expect, it } from "vitest";
import { APP_NAME, APP_DESCRIPTION } from "$lib/brand";
import { applyPageMeta, metaFromRoute } from "./meta";

describe("seo meta", () => {
  beforeEach(() => {
    document.head.innerHTML = "";
    document.title = "";
  });

  it("builds route meta with fallbacks", () => {
    const meta = metaFromRoute(
      { path: "/music/albums", title: "Albums", description: "Browse albums." },
      "/music/albums",
    );
    expect(meta.title).toBe("Albums");
    expect(meta.description).toBe("Browse albums.");
    expect(meta.path).toBe("/music/albums");
  });

  it("applies title description and open graph tags", () => {
    applyPageMeta({
      title: "Albums",
      description: "Browse albums in the library.",
      path: "/music/albums",
      image: "/og.png",
    });

    expect(document.title).toBe(`Albums · ${APP_NAME}`);
    expect(
      document.head
        .querySelector('meta[name="description"]')
        ?.getAttribute("content"),
    ).toBe("Browse albums in the library.");
    expect(
      document.head
        .querySelector('meta[property="og:title"]')
        ?.getAttribute("content"),
    ).toBe(`Albums · ${APP_NAME}`);
    expect(
      document.head
        .querySelector('meta[property="og:image"]')
        ?.getAttribute("content"),
    ).toContain("/og.png");
    expect(
      document.head
        .querySelector('link[rel="canonical"]')
        ?.getAttribute("href"),
    ).toContain("/music/albums");
  });

  it("falls back to the app description", () => {
    applyPageMeta({ title: APP_NAME });
    expect(
      document.head
        .querySelector('meta[name="description"]')
        ?.getAttribute("content"),
    ).toBe(APP_DESCRIPTION);
  });
});

File: internal/seo/meta_test.go
// SPDX-FileCopyrightText: 2026 Quad4 Software
// SPDX-License-Identifier: Apache-2.0

package seo

import (
	"strings"
	"testing"
)

func TestForPath(t *testing.T) {
	page := ForPath("/music/albums")
	if page.Title != "Albums" {
		t.Fatalf("title %q", page.Title)
	}
	if !strings.Contains(page.Description, "albums") {
		t.Fatalf("description %q", page.Description)
	}

	settings := ForPath("/settings/playback")
	if settings.Index {
		t.Fatal("settings should be noindex")
	}

	unknown := ForPath("/nope")
	if unknown.Title == "" || unknown.Description == "" {
		t.Fatal("unknown path should still have defaults")
	}
}

func TestInject(t *testing.T) {
	raw := []byte(`<!doctype html><html><head>
<title>Old</title>
<meta name="description" content="old" />
</head><body></body></html>`)
	page := Page{
		Title:       "Albums",
		Description: "Browse albums",
		Image:       "/og.png",
		Type:        "website",
		Index:       true,
	}
	out := string(Inject(raw, page, "https://music.example", "/music/albums"))
	checks := []string{
		"<title>Albums · Melovian</title>",
		`content="Browse albums"`,
		`property="og:url" content="https://music.example/music/albums"`,
		`property="og:image" content="https://music.example/og.png"`,
		`rel="canonical" href="https://music.example/music/albums"`,
		`name="twitter:card" content="summary_large_image"`,
	}
	for _, want := range checks {
		if !strings.Contains(out, want) {
			t.Fatalf("missing %q in:\n%s", want, out)
		}
	}
}
